base/compiler/typeutils.jl

```diff
@@ -70,6 +70,12 @@ function typesubtract(@nospecialize(a), @nospecialize(b))
     if isa(a, Union)
         return Union{typesubtract(a.a, b),
                      typesubtract(a.b, b)}
+    elseif a isa DataType && b isa DataType && a.name === b.name === Tuple.name &&
+            length(a.types) == length(b.types)
```

Comment:
It can be better if you write this as an `if` statement diamond, since inference currently can only propagate the last `isa` constraint on a line
```suggestion
    elseif a isa DataType
        if b isa DataType
            if a.name === b.name === Tuple.name && length(a.types) == length(b.types)
```

Reply:
> It can be better if you write this as an if statement diamond, since inference currently can only propagate the last isa constraint on a line

@vtjnash does this limit still hold true ? I tried the following snippet 
```julia
code_typed((Any,Any); optimize = false) do a, b
    if isa(a, Int) && isa(b, Int) && (a > 0 || b > 0)
        sin(a)
    end
end |> first |> first
```
and found `a` on line 3 is inferred as `Int` as expected:
```julia
CodeInfo(
1 ─ %1  = (a isa TypeProfiler.Int)::Bool
└──       goto #8 if not %1
2 ─ %3  = (b isa TypeProfiler.Int)::Bool
└──       goto #8 if not %3
3 ─ %5  = (a::Int64 > 0)::Bool
└──       goto #5 if not %5
4 ─       (@_4 = %5)::Bool
└──       goto #6
5 ─       (@_4 = b::Int64 > 0)::Bool
6 ┄       goto #8 if not @_4
7 ─ %11 = TypeProfiler.sin(a::Int64)::Float64
└──       return %11
8 ┄       return nothing
)
```

Reply:
No, the frontend usually rewrites this syntax now to form the diamond automatically

Reply:
nice, could you please tell me which PR/commit I should look at to see the frontend changes ?

Reply:
https://github.com/JuliaLang/julia/pull/36355

Reply:
thx